prov/verbs/src/verbs_domain_xrc.c

```diff
@@ -0,0 +1,576 @@
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#include "config.h"
+#include "fi_verbs.h"
+
+#ifdef INCLUDE_VERBS_XRC
+
+/* Domain XRC INI QP RBTree key */
+struct fi_ibv_ini_conn_key {
+	struct sockaddr		*addr;
+	struct fi_ibv_cq	*tx_cq;
+};
+
+struct fi_ibv_domain *fi_ibv_msg_ep_to_domain(struct fi_ibv_ep *ep)
+{
+	struct fi_ibv_domain *domain = NULL;
+
+	if (ep->util_ep.tx_cq)
+		domain = container_of(ep->util_ep.tx_cq->domain,
+				      struct fi_ibv_domain, util_domain);
+	else if (ep->util_ep.rx_cq)
+		domain = container_of(ep->util_ep.rx_cq->domain,
+				      struct fi_ibv_domain, util_domain);
+
+	/* Only valid to call this function after CQ(s) are bound */
+	assert(domain);
+	return domain;
+}
+
+/*
+ * This routine is a work around that creates a QP for the only purpose of
+ * reserving the QP number. The QP is not transitioned out of the RESET state.
+ */
+struct ibv_qp *fi_ibv_reserve_qpn(struct fi_ibv_ep *ep)
+{
+	struct fi_ibv_domain *domain = fi_ibv_msg_ep_to_domain(ep);
+	struct fi_ibv_cq *cq = container_of(ep->util_ep.tx_cq,
+					    struct fi_ibv_cq, util_cq);
+	struct ibv_qp *qp;
+	struct ibv_qp_init_attr attr = { 0 };
+	int ret;
+
+	/* Limit library allocated resources and do not INIT QP */
+	attr.cap.max_send_wr = 1;
+	attr.cap.max_send_sge = 1;
+	attr.cap.max_recv_wr = 0;
+	attr.cap.max_recv_sge = 0;
+	attr.cap.max_inline_data = 0;
+	attr.send_cq = cq->cq;
+	attr.recv_cq = cq->cq;
+	attr.qp_type = IBV_QPT_RC;
+
+	qp = ibv_create_qp(domain->pd, &attr);
+	if (!qp) {
+		ret = -errno;
+		VERBS_INFO_ERRNO(FI_LOG_EP_CTRL,
+				 "Reservation QP create failed", ret);
+		return NULL;
+	}
+	return qp;
+}
```

Comment:
Clearly it would preferable not to, but in my opinion the benefit of using the RDMA CM for all connections (physical or shared) between endpoints out weighs allocating a minimal QP, leaving it in the reset state, and deleting it once the connection completes. This prevents the possibility of collisions with actual HCA QP number space.